res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js

```diff
@@ -0,0 +1,978 @@
+        var timeDelta = deltas[1];
+
+        var velocity = (tickDelta / timeDelta) / 3;
+        engine.setValue(field.group, "jog", velocity);
+        if (engine.getValue(field.group, "scratch2_enable")) {
+            engine.scratchTick(deckNumber, tickDelta);
+        }
+    }
+};
+
+TraktorS2MK3.wheelDeltas = function (group, value) {
+    // When the wheel is touched, four bytes change, but only the first behaves predictably.
+    // It looks like the wheel is 1024 ticks per revolution.
+    var tickval = value & 0xFF;
+    var timeval = value >>> 16;
+    var prevTick = 0;
+    var prevTime = 0;
+
+    if (group[8] === "1" || group[8] === "3") {
+        prevTick = this.lastTickVal[0];
+        prevTime = this.lastTickTime[0];
+        this.lastTickVal[0] = tickval;
+        this.lastTickTime[0] = timeval;
+    } else {
+        prevTick = this.lastTickVal[1];
+        prevTime = this.lastTickTime[1];
+        this.lastTickVal[1] = tickval;
+        this.lastTickTime[1] = timeval;
+    }
+
+    if (prevTime > timeval) {
+        // We looped around.  Adjust current time so that subtraction works.
+        timeval += 0x10000;
+    }
+    var timeDelta = timeval - prevTime;
+    if (timeDelta === 0) {
+        // Spinning too fast to detect speed!  By not dividing we are guessing it took 1ms.
+        timeDelta = 1;
+    }
+
+    var tickDelta = 0;
+    if (prevTick >= 200 && tickval <= 100) {
+        tickDelta = tickval + 256 - prevTick;
+    } else if (prevTick <= 100 && tickval >= 200) {
+        tickDelta = tickval - prevTick - 256;
+    } else {
+        tickDelta = tickval - prevTick;
+    }
+
+    return [tickDelta, timeDelta];
+};
+
+TraktorS2MK3.fxHandler = function (field) {
+    if (field.value === 0) {
+        return;
+    }
+
+    var fxNumber = parseInt(field.id[field.id.length - 1]);
+    var group = "[EffectRack1_EffectUnit" + fxNumber + "]";
+
+    // Toggle effect unit
+    TraktorS2MK3.fxButtonState[fxNumber] = !TraktorS2MK3.fxButtonState[fxNumber];
+
+    engine.setValue(group, "group_[Channel1]_enable", TraktorS2MK3.fxButtonState[fxNumber]);
+    engine.setValue(group, "group_[Channel2]_enable", TraktorS2MK3.fxButtonState[fxNumber]);
+    TraktorS2MK3.outputHandler(TraktorS2MK3.fxButtonState[fxNumber], field.group, "fxButton" + fxNumber);
+};
+
+TraktorS2MK3.reverseHandler = function (field) {
+    if (TraktorS2MK3.shiftPressed[field.group]) {
+        engine.setValue(field.group, "reverseroll", field.value);
+    } else {
+        engine.setValue(field.group, "reverse", field.value);
+    }
+
+    TraktorS2MK3.outputHandler(field.value, field.group, "reverse");
+};
+
+TraktorS2MK3.flxHandler = function (field) {
+    if (field.value === 0) {
+        return;
+    }
+
+    var slip = engine.getValue(field.group, "slip_enabled");
+    engine.setValue(field.group, "slip_enabled", !slip);
+    TraktorS2MK3.outputHandler(!slip, field.group, "flx");
+};
+
+TraktorS2MK3.gridHandler = function (field) {
```

Comment:
Rename to `beatgridHandler`. At first I thought this referred to the pad grid

Reply:
Done in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468